src/ConstrainedPhotometryModel.cc

```diff
@@ -0,0 +1,208 @@
+        CcdImage const &ccdImage) const {
+    auto oldPhotoCalib = ccdImage.getPhotoCalib();
+    auto detector = ccdImage.getDetector();
+    auto ccdBBox = detector->getBBox();
+    ChipVisitPhotometryMapping *mapping =
+            dynamic_cast<ChipVisitPhotometryMapping *>(this->findMapping(ccdImage, "toPhotoCalib"));
+    // There should be no way in which we can get to this point and not have a ChipVisitMapping,
+    // so blow up if we don't.
+    assert(mapping != nullptr);
+    auto pixToFocal = detector->getTransform(afw::cameraGeom::PIXELS, afw::cameraGeom::FOCAL_PLANE);
+    // We know it's a Chebyshev transfo because we created it as such, so blow up if it's not.
+    auto visitTransfo =
+            std::dynamic_pointer_cast<PhotometryTransfoChebyshev>(mapping->getVisitMapping()->getTransfo());
+    assert(visitTransfo != nullptr);
+    auto focalBBox = visitTransfo->getBBox();
+
+    // Unravel our chebyshev coefficients to build an astshim::ChebyMap.
+    auto coeff_f = toChebyMapCoeffs(
+            std::dynamic_pointer_cast<PhotometryTransfoChebyshev>(mapping->getVisitMapping()->getTransfo()));
+    // Bounds are the bbox
+    std::vector<double> lowerBound = {focalBBox.getMinX(), focalBBox.getMinY()};
+    std::vector<double> upperBound = {focalBBox.getMaxX(), focalBBox.getMaxY()};
+    ast::ChebyMap chebyMap(coeff_f, 1, lowerBound, upperBound);
+
+    // The chip part is easy: zoom map with the single value as the "zoom" factor.
+    ast::ZoomMap zoomMap(1, mapping->getChipMapping()->getParameters()[0]);
+
+    // Now stitch them all together.
+    auto transform =
+            afw::geom::TransformPoint2ToGeneric(pixToFocal->getFrameSet()->then(chebyMap).then(zoomMap));
+    // NOTE: TransformBoundedField does not yet implement mean(), so we have to compute it here.
+    double mean = mapping->getChipMapping()->getParameters()[0] * visitTransfo->mean();
+    auto boundedField = std::make_shared<afw::math::TransformBoundedField>(ccdBBox, transform);
+    return std::make_shared<afw::image::PhotoCalib>(mean, oldPhotoCalib->getCalibrationErr(), boundedField,
+                                                    false);
+}
+
+void ConstrainedPhotometryModel::dump(std::ostream &stream) const {
+    for (auto &i : _chipMap) {
+        i.second->dump(stream);
+        stream << std::endl;
+    }
+    stream << std::endl;
+    for (auto &i : _visitMap) {
+        i.second->dump(stream);
+        stream << std::endl;
+    }
+}
+
+PhotometryMappingBase *ConstrainedPhotometryModel::findMapping(CcdImage const &ccdImage,
+                                                               std::string name) const {
+    auto i = _myMap.find(ccdImage.getHashKey());
+    if (i == _myMap.end())
+        throw LSST_EXCEPT(
+                pex::exceptions::InvalidParameterError,
+                "ConstrainedPhotometryModel::" + name + ", cannot find CcdImage " + ccdImage.getName());
```

Comment:
Why  is `name` a parameter that only occurs in the error string? This seems like a recipe for wrong information.

Reply:
I wrote it that way so that the error can describe what the caller was, since this is used in many internal methods. Should I just rely on `throw`'s stack trace for that purpose?

This is the declaration in the base class:

```C++
/// Return a pointer to the mapping associated with this ccdImage. name is for describing error messages.
virtual PhotometryMappingBase *findMapping(CcdImage const &ccdImage, std::string name) const = 0;
```

Reply:
Yeah, I think the stack trace is fine for that. If it is the name of the class you could add a `getName` virtual function and call that instead.

Reply:
it's not the name of the class, but the name of the calling function. But I've removed `name` since you think the stack trace is enough.